### ISSUE TITLE
Support Bluetooth Transactions

### ISSUE THREAD

**Issue by throughnothing** [voisine/breadwallet#152]
Would be nice to support transacting with another phone/device over Bluetooth.  I'm not sure if there's a real standard around this yet, but other wallets have done it, such as AirBitz: https://airbitz.co/go/pay-bluetooth-now-android/

> The protocol for payments via BLE is an open standard currently being reviewed by the Bitcoin core  developer community. Airbitz looks forward to collaboration with other developers to help improve and secure the protocol to help drive its adoption in other wallets and payment processors.

Would be great to inter-operate over bluetooth with other (non-breadwallet) software and (non-iPhone) devices.

**Comment by voisine** [comment]
We'll be following the development. As it stands there are some problems with their design both for usability and security, but we definitely want to enable Bluetooth payments so bitcoin can be used outside of cell service areas.

**Comment by throughnothing** [comment]
Yes, after reading the discussion, it seems like they're using the `bitcoin:` url scheme, instead of Payment Protocol protobufs/objects, which seems silly.

I guess there's no point implementing anything until the details are ironed out, but it'll be great to have once the spec is there.

**Comment by nshopik** [comment]
What happends to #92 ?

**Comment by voisine** [comment]
#92 is tied to apple's proprietary multipeer connectivity api, but it would be cool to support both standard bluetooth and multipeer if available.

There's also an issue raised on the dev list where an attacker can jam and replace a legitimate bluetooth signal instead of just duplicating it, but I suppose it no worse than copying an address to the clipboard or scanning from a desktop computer or getting it email, all of which are pretty insecure. We'll have to migrate away from all of those toward more secure options over time.

**Comment by throughnothing** [comment]
Seems like payment protocol will alleviate all of those things (even when done over email, bluetooth, whatever), as long as you've done a transaction with that person before, so can verify their cert/key/whatever.

**Comment by mikehearn** [comment]
You could aim for compatibility with the Schildbach wallet. The protocol used there is quite simple and it's in use in the wild already.

**Comment by keo** [comment]
:+1:

**Comment by dabura667** [comment]
BIP70 and NFC / Bluetooth is the way of the future... Apple Pay, but with secure digital cash.

If there's no standard, breadwallet should lead the way to decide on a BIP for Bluetooth / NFC communication between wallets.

**Comment by keo** [comment]
Here's Andreas' wiki page on his implementation: https://github.com/schildbach/bitcoin-wallet/wiki/Payment-Requests

It is a good read and I like this direction better than the one Airbitz took.

**Comment by schildbach** [comment]
Yes, it would be great if Breadwallet would support the Bluetooth payment protocol. Bitcoin Wallet users use it a lot, and like the fact that transactions arrive immediately even if their WLAN is dodgy.

**Comment by Logicwax** [comment]
How about starting with leveraging BLE and "suggested apps" on the lockscreen?   It is to my understanding that a BLE device MAJOR (as opposed to MINOR) advertisement code can be set to a specific application, wherein whenever that MAJOR code is heard the application icon will be presented on the bottom left-hand side of the lockscreen for quick access.   It would be a nice feature, if a bunch of developers could agree on a MAJOR code (for example 0xB17C01N....or whatever) that would act as a recognized "wake-up wallet!" flag.   Going farther, a quick swipe from the lockscreen would load up Breadwallet with the camera-scanner already running (like how notification center widget works).

Then all POS terminals (or other Breadwallets even) would just simply burp out those MAJOR codes in their advertisement messages, so that users could rapidly jump into the application whenever the opportunity to pay with bitcoin arises.

**Comment by voisine** [comment]
@Logicwax that is really cool idea... this sounds like something that could be written up in a BIP. We have a good relationship with coinbase and bitpay in the US and coincheck in japan among others, so we can approach them about supporting it in their merchant POS terminal apps.

It looks like on iOS it's possible to support iBeacons using the CoreLocation framework. For security reasons, we can also require that any payment requests over BLE use some form of payee identity certification, either via a QR scan or X.509 or something else that comes into popular use later. We need to defend against an attacker creating a dummy beacon while standing behind you in the checkout line.

**Comment by schildbach** [comment]
Payment Requests can already be X.509 signed, so that's straight forward.

**Comment by voisine** [comment]
@schildbach I was thinking with a QR scan, we can accept an uncertified payment request broadcast over BLE at the time of the scan, if the payment address matches the QR.

**Comment by schildbach** [comment]
Hmmm, if you scan via QR, what use would BLE then be? The current payment protocol is already implemented for standard Bluetooth.

**Comment by voisine** [comment]
It would be useful for payment protocol additions like receipts and refund addresses. Using the bitcoin address to tie the request to the QR allows the QR to be no larger than it is already. The problem with using BLE without a QR scan or some other certification method as a trust anchor is that an attacker in close proximity can trick you into paying the wrong address. We can also support BLE requests without a QR scan if they are certified another way, such as X.509

**Comment by schildbach** [comment]
Refund addresses is already in BIP70. Receipts could be added as an extension to BIP70, but that would not change the way how it's handled via BT. The trust anchor is currently either a QR code or NFC, X.509 is only used for fetching a name.

Maybe I'll ask this way: What's the usecase exactly? If you start by scanning a QR code, you can just as well use existing protocols. Almost all BLE devices also can do classic Bluetooth.

**Comment by voisine** [comment]
Yes, I'm talking about getting the advantages of BIP70 using BLE.

If you are scanning a QR, you might not have internet access to fetch the payment request over http, and the classic bluetooth extension makes the QR larger and more difficult to scan. The wallet can just see if there are any requests being broadcast over BLE and if one matches the address of the QR, you've established an authenticated BLE communication channel with the intended payee, even if the payment request itself isn't certified using something like X.509, so it could work with person-to-person payments.

**Comment by Logicwax** [comment]
@voisine:  actually, I have been working on my own merchant cellular-based POS terminals, and will be going public with them soon (hence, why in the other thread, I'm irritated by breadwallet connection issues lately.....because I've been using breadwallet for all my beta testing).   The terminals use QR-codes.  

No need to deal with security of payee identity just yet.   I'm simply proposing a mechanism for putting breadwallet on the lockscreen, ready at the simple flick of a finger to launch right into camera mode, by using BLE as an advertisement.   Various retail outlets are already using BLE for this exact purpose.   


As for security of payee identity, I always wondered......why not just the payor and payee perform a Diffie-Hellman to come up with a shared secret.   Then use that shared secret to select a basic color, or one of 20 icon images.   If you see red (or a picture of a cow) on your phone, and on the POS terminal, then you know you are talking to the right payee.  If you see blue (or a picture of a smiley face) on the POS terminal, then you know there's an attacker.   but I digress......

My main point is, using BLE to get Breadwallet ready on the lockscreen.  I want to implement this on my terminals ASAP.   I just need Breadwallet to register to be woken up via locationservices from some agreed-upon MAJOR code.

**Comment by schildbach** [comment]
@Logicwax As soon as you come up with something, let me know. I'm interested in implementing that simple feature for Android too. It might involve supporting a second beacon standard though. Android has committed to Eddystone. But I guess your POS could broadcast two beacons simultaneously as well?

**Comment by schildbach** [comment]
@voisine So you're suggesting that instead of a QR code with address, amount and r=bt:<mac> parameter, the QR code contains just address and amount and the wallet prophylactically scans the air for a BTLE message/beacon that contains the full URL and matches address and amount? If found, at that point BIP70 would continue as usual, so it would fetch a payment request from either HTTP(S) or a Bluetooth socket depending on the contant of the r= parameter.

**Comment by voisine** [comment]
@Logicwax I like the idea of a beacon to launch the app as a first step. I would just like to do it in a way that we can latter improve it so that when talking to an X.509 certified merchant terminal, you can skip the scan entirely, you just go straight to a payment authorization dialog. Then the next improvement would be to communicate the payment over BLE so it works even if the customer doesn't have internet access. Then thirdly enabling the QR scan to be the trust anchor for non-certified BIP70 payment requests received over BLE to ensure you are paying the intended person.

@schildbach I wasn't actually thinking of putting the url in the beacon itself, but that's quite a good idea. Then you don't have to actually connect to see if the address matches the QR scan. Perhaps the QR would have a normal https r= argument, and the BLE beacon url would have something like r=BLE indicating that the BLE source can be connected to for a BIP70 exchange. (not exactly sure how beacons work, can you safely establish a connection with the specific BLE device emitting a beacon?)

**Comment by schildbach** [comment]
@voisine Sounds great. Rather than r=BLE I'd suggest using r=bt:<mac> though.
